.github/workflows/validate-migration.yaml

```diff
@@ -5,49 +5,107 @@ on: [ push ]
+          - valgrindHelgrind # https://valgrind.org/info/tools.html#helgrind
           - noSanitizer
-        include:
-          - sanitizer: addressSanitizer
-            CMAKE_CXX_FLAGS: "-fsanitize=address -g"
-          - sanitizer: threadSanitizer
-            CMAKE_CXX_FLAGS: "-fsanitize=thread -g"
-          #          disabled until we figure out how to build with clang in github actions
-          #          - sanitizer: memorySanitizer
-          #            CMAKE_CXX_FLAGS: "-fsanitize=memory -g"
-          - sanitizer: leakSanitizer
-            CMAKE_CXX_FLAGS: "-fsanitize=leak -g"
-          - sanitizer: undefinedSanitizer
-            CMAKE_CXX_FLAGS: "-fsanitize=undefined -g"
-          - sanitizer: noSanitizer
-            CMAKE_CXX_FLAGS: ""
 
     runs-on: ubuntu-latest
 
     steps:
+      - name: "Prepare environment for specific tool"
+        run: |
+          # Set environment variables for future steps based on the tool being used. We can't refer to the matrix variables in the env section
+          # of the job, so we have to do it here.
+          #
+          # See https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable 
+          # for more information on the method to set environment variables.
+          
+          bash <<EOF
+            set -euxo pipefail
+          
+            case ${{ matrix.tool }} in
+            
+            noSanitizer)
+              echo "USE_VALGRIND=false" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=" >> $GITHUB_ENV
+            ;;
+          
+            undefinedSanitizer)
+              echo "USE_VALGRIND=false" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=-fsanitize=undefined -fno-sanitize-recover=all -g" >> $GITHUB_ENV
+              echo "OPENSSL_FLAGS=-fsanitize=undefined -fno-sanitize-recover=all -g" >> $GITHUB_ENV
+            ;;
+          
+            addressSanitizer)
+              echo "USE_VALGRIND=false" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=-fsanitize=address -fno-sanitize-recover=all -g" >> $GITHUB_ENV
+              echo "OPENSSL_FLAGS=-fsanitize=address -fno-sanitize-recover=all -g" >> $GITHUB_ENV
+            ;;
+          
+            threadSanitizer)
+              echo "USE_VALGRIND=false" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=-fsanitize=thread -fno-sanitize-recover=all -g" >> $GITHUB_ENV
+              echo "OPENSSL_FLAGS=-fsanitize=thread -fno-sanitize-recover=all -g" >> $GITHUB_ENV
+            ;;
+          
+            valgrindMemcheck)
+              echo "USE_VALGRIND=true" >> $GITHUB_ENV
+              echo "CMAKE_CXX_FLAGS=-g" >> $GITHUB_ENV
+              echo "VALGRIND_FLAGS=--tool=memcheck --leak-check=full --show-leak-kinds=all --track-origins=yes" >> $GITHUB_ENV
+            ;;
+          
+            valgrindHelgrind)
+              echo "USE_VALGRIND=true" >> $GITHUB_ENV
```

Comment:
Suggestion: Remove this `USE_VALGRIND` variable and instead check if `VALGRIND_FLAGS` is defined or not at the point of use.